plugin/pkg/auth/authenticator/password/keystone/keystone.go

```diff
@@ -45,18 +50,74 @@ func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticatePassword(usernam
 		glog.Info("Failed: Starting openstack authenticate client")
 		return nil, false, errors.New("Failed to authenticate")
 	}
-
+//FIXME user_id, not user_name...
 	return &user.DefaultInfo{Name: username}, true, nil
 }
 
-// NewKeystoneAuthenticator returns a password authenticator that validates credentials using openstack keystone
-func NewKeystoneAuthenticator(authURL string) (*KeystoneAuthenticator, error) {
-	if !strings.HasPrefix(authURL, "https") {
-		return nil, errors.New("Auth URL should be secure and start with https")
+func (keystoneAuthenticator *KeystoneAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+//FIXME kfox1111: To enhance performance, cache admin token/client.
+	cfg, provider, err := loadConfig(keystoneAuthenticator.configPath)
+	if err != nil {
+		return nil, false, errors.New("Internal error getting the Keystone provider")
+	}
+
+	client, err := openstack.NewIdentityAdminV3(provider, gophercloud.EndpointOpts{
+		Region: cfg.Global.Region,
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	response, err := client.Request("GET", client.ServiceURL("auth", "tokens"), gophercloud.RequestOpts{
+	        MoreHeaders: map[string]string{"X-Subject-Token": value},
+	        OkCodes:     []int{200},
+	})
+	if err != nil {
+		return nil, false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.V(3).Infof("Cannot get HTTP response body from keystone token validate: %v", err)
+		return nil, false, err
+	}
+	obj := struct{ Token struct { User struct { Id string }}}{}
+	err = json.Unmarshal(bodyBytes, &obj)
+	if err != nil {
+		return nil, false, err
+	}
+	return &user.DefaultInfo{Name: obj.Token.User.Id}, response.StatusCode == 200, nil
```

Comment:
> In this case, Users should just be restricted to Users that have a specified Role on a specified Project so that only Users of that Project can perform actions.

If the level of granularity you're looking for is just "can this user use this kubernetes installation" without regard for what specific action they're taking against the kubernetes API, I *guess* the authenticator could make that decision, and users that didn't meet the requirements wouldn't even be considered users (they'd get a 401 instead of a 403 error).

If you're looking to make authorization decisions based on the kubernetes namespace, or resources, or action being performed, etc, that information is available to the authorizer interface (along with the `user.Info` returned by the authenticator). If you wanted to pair a keystone authenticator and authorizer, you could plumb user domain/tenant/role information from the authenticator to the authorizer in the `Extra` user attributes (added recently in https://github.com/kubernetes/kubernetes/pull/23574/files#diff-4f723ba6c5ced0845bfaa4ff3a5ad25b).